### PR TITLE
ci(coverage-gate): stop swallowing vitest coverage-threshold failures

### LINKED ISSUE #8598: [F06] Coverage threshold failures are silently swallowed by the vitest#3077 exit-code workaround
**Audit finding F06** · 🟠 **medium** · dimension `ci` (testing) · phase **P1**

Part of epic #8591. Source: SpawnForge Security & Testing Audit (2026-05-30).

**File:** `.github/workflows/quality-gates.yml:178-187`

The test-web job wraps `vitest run --coverage` and post-processes the exit code: if exit != 0, it only propagates the failure when the output contains a line matching `Test Files.*failed`; otherwise it prints a warning and `exit 0`. I verified in the installed vitest 4.x source (node_modules/vitest/dist/chunks/coverage.DM_a_rWm.js:879-893) that a coverage-threshold violation sets `process.exitCode = 1` and logs `ERROR: Coverage for statements (X%) does not meet global threshold (Y%)`. Crucially, when all tests pass but a threshold is missed, the reporter's `Test Files` summary line (index.UpGiHP7g.js:3023 + getStateString at :3059) only emits the word `failed` when `entry.failed > 0`. With zero failed tests it reads `Test Files  N passed (N)` — no `failed` substring. The grep at line 181 therefore matches nothing, the else-branch runs, and the job exits 0 despite the coverage gate failing. The identical logic is duplicated in cd.yml:180-187.

## Impact
The 75/65/70/77 coverage thresholds in web/vitest.config.ts are not actually enforced. A PR that drops coverage below threshold (e.g. by adding large untested modules) passes the Web Tests gate. The project's coverage-as-a-gate guarantee — explicitly called out as a CI invariant in CLAUDE.md — is silently broken. Only real test failures block; any non-test-failure non-zero exit (including the coverage gate) is treated as a benign open-handle artifact.

## Evidence
quality-gates.yml:180 `if sed 's/\x1b\[[0-9;]*m//g' /tmp/vitest-output.txt | grep -q "Test Files.*failed"; then` ... else `exit 0`. vitest source coverage.DM_a_rWm.js:885 `errorMessage = \`ERROR: Coverage for ${thresholdKey} (${coverage}%) does not meet ... threshold\`` with `process.exitCode = 1` at :879.

## Recommended fix
Distinguish coverage-threshold failures from open-handle hangs. After the run, grep the captured output for `ERROR: Coverage for` / `does not meet .* threshold` and propagate a non-zero exit when present. Better: drop the timeout/grep heuristic entirely and run coverage in a separate `vitest run --coverage` invocation with `--reporter=json` + an explicit threshold check, or use `--coverage.thresholds.autoUpdate=false` with `--bail`, so the exit code is trusted. Reserve the exit-124 special-case strictly for the literal 124 timeout code (already handled at line 174).

---
Full report & phased plan: `docs/audits/2026-05-30-security-testing-audit.md`

### PR DESCRIPTION
**Pull request by Tristan578**
## Summary
- The vitest#3077 exit-code workaround (treating `timeout` exit 124 as a warning) was also swallowing real coverage-threshold failures: a run that failed thresholds could exit through the workaround path and report green. Coverage gates in `quality-gates.yml` and `cd.yml` were therefore advisory in practice ([F06]).
- Extract the exit-code decision into `scripts/check-vitest-exit.sh`: exit 124 (jsdom open-handle hang after tests complete) stays a warning, but any output matching vitest's coverage-threshold failure forms — including the max-uncovered form — fails the job regardless of exit code. ANSI escape sequences are stripped before matching (covered by a split-token mid-word test case).
- TDD: `scripts/__tests__/check-vitest-exit.test.sh` covers the boundary cases (plain threshold form, max-uncovered form, ANSI-split tokens, timeout-only exit, malformed input fail-safe).
- Wired into `ci.yml`, `quality-gates.yml`, and `cd.yml` so all three coverage paths share one verifier.

Closes #8598

## Test plan
- [x] `shellcheck` clean on `scripts/check-vitest-exit.sh` + test suite
- [x] `bash scripts/__tests__/check-vitest-exit.test.sh` — PASS
- [x] `actionlint` on all three workflows — no new findings vs origin/main (net −6: branch fixes 6 pre-existing SC2086s in rewritten regions; remaining legacy findings tracked in #8719)
- [ ] CI green on this PR

### CODE REVIEW

File: scripts/check-vitest-exit.sh [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+case "$EXIT_CODE" in
+  ''|*[!0-9]*) usage ;;
+esac
+
+# 124 = `timeout` killed a process that hung during cleanup after tests already
+# completed (vitest#3077). Treat as a warning, not a failure.
+if [ "$EXIT_CODE" -eq 124 ]; then
+  echo "::warning::vitest process was killed after timeout — likely hung during cleanup (vitest#3077)"
+  exit 0
+fi
```
[the remark this answers is not in the document]
Comment (Tristan578) [reply scripts/check-vitest-exit.sh]: Fixed in ef4f2389 — the early exit-124 swallow is deleted, so a timed-out run now flows through the full evidence pipeline: fail-closed missing/empty-evidence check, ANSI stripping, failed-test marker scan, and coverage-threshold adjudication. Five regression cases added in scripts/__tests__/check-vitest-exit.test.sh (124+threshold miss, 124+failed test files, 124+ANSI-colored markers, 124+clean passing evidence stays green, 124+missing evidence fails closed). Follow-up f7e6d11d corrects the gate's comments to state the exact guarantee the coverage marker provides, verified against the pinned @vitest/coverage-v8 4.1.7 source.
